oshmem/mca/spml/ucx/spml_ucx.c

```diff
@@ -104,11 +104,69 @@ int mca_spml_ucx_enable(bool enable)
     return OSHMEM_SUCCESS;
 }
 
+/* add a new mkey and update update the mkeys_cnt */
+int mca_spml_ucx_ep_mkey_add(ucp_peer_t *ucp_peer, int index)
+{
+    int old_size = ucp_peer->mkeys_cnt;
+    if (MCA_MEMHEAP_MAX_SEGMENTS <= ucp_peer->mkeys_cnt) {
+        SPML_UCX_ERROR("Failed to get new mkey for segment: max number (%d) of segment descriptor is exhausted",
+                       MCA_MEMHEAP_MAX_SEGMENTS);
+        return OSHMEM_ERROR;
+    }
+    /* Allocate an array to hold the pointers to the ucx_cached_mkey */
+    if (index >= ucp_peer->mkeys_cnt){
+        ucp_peer->mkeys_cnt = index + 1;
+        ucp_peer->mkeys = realloc(ucp_peer->mkeys, sizeof(spml_ucx_cached_mkey_t *) * ucp_peer->mkeys_cnt);
+        if (NULL == ucp_peer->mkeys) {
+            SPML_UCX_ERROR("Failed to obtain new mkey: OOM - failed to expand the descriptor buffer");
+            return OSHMEM_ERR_OUT_OF_RESOURCE;
+        }
+        /* NOTE: release code checks for the rkey != NULL as a sign of used element:
+        Account for the following scenario below by zero'ing the unused elements:
+        |MKEY1|00000|MKEY2|??????|NEW-MKEY|
+        |<--- old_size -->|
+        */
+        memset(ucp_peer->mkeys + old_size, 0, (index + 1 - old_size) * sizeof(spml_ucx_cached_mkey_t *));
+    } else {
+        assert(NULL == ucp_peer->mkeys[index]);
+    }
+    
+    ucp_peer->mkeys[index] = (spml_ucx_cached_mkey_t *) malloc(sizeof(spml_ucx_cached_mkey_t));
+    if (NULL == ucp_peer->mkeys[index]) {
+        SPML_UCX_ERROR("Failed to obtain new ucx_cached_mkey: OOM - failed to expand the descriptor buffer");
+        return OSHMEM_ERR_OUT_OF_RESOURCE;
+    }
+    return OSHMEM_SUCCESS;
+}
+
+/* Release individual mkeys */
+void mca_spml_ucx_ep_mkey_release(ucp_peer_t *ucp_peer, int segno)
+{
+    if ((ucp_peer->mkeys_cnt >= segno) || (segno < 0))
+    {
```

Comment:
Slould it be "less or equal?
```suggestion
    if ((ucp_peer->mkeys_cnt <= segno) || (segno < 0)) {
```